[WIP] 131 refcell world

Vrixyz opened the pull request:
**This Pull Request is NOT READY to merge now !**


Referencing https://github.com/sebcrozet/nphysics/issues/131

## Current state
I succeeded in compiling:  
- balls2d official nphysics example, which is working fine.
- [my target repository](https://github.com/Vrixyz/testbed-callback/tree/1-arc-mutex) compiled, but I must have some deadlock in there, I'll figure it out when I come back to it.

I'm opening the pull request to show my progress, and discuss about the shortcomings.

## Attempts
### Very first attempt
Within Testbed, I tried to replace `World` by `&mut World`, no chance in compiling ever after.

### Second attempt
Within Testbed, I tried to add a "WorldOwner", which objective was to own a reference to a &mut World, then have a `retrieve` function to access it.

I lost myself again in [complex lifetime annotations](https://github.com/rust-lang/rust/issues/53343).

### Third attempt
I started small, read [some articles](https://doc.rust-lang.org/book/second-edition/ch15-05-interior-mutability.html) (again)
Then, I used `RefCell<World>` to store a mutable world inside `Testbed`.
Once compiling and at least 1 official example working, I used `Rc<RefCell<World>`.

Balls2d compiled successfully, but not [my reproduction repository](https://github.com/Vrixyz/testbed-callback/tree/1-arc-mutex) : indeed, specs::System needs its resources to be `sync::Send` to be sent beetween threads.

So I tried to `Arc<RefCell<World>>`, but I quickly realized I had no choice but to use `Arc<RefCell<World>>`, since `RefCell` is not `sync::Send`.

## Shortcomings

- I had to remove <nphysics::World> argument from callback : as it is shared, it's useless to pass it there.
- Using Arc<RefCell<World>> is verbose and useless in most usecases, so we're sacrificing quite a lot readability, and @sebcrozet vision for nphysics is simplicity...

### Improvements

Once my use case working, I'll see if I can make both ways compatibles:
- First using a raw `World`, Testbed taking the only ownership to it (current behaviour)
- Second using a complex Arc<Mutex<World>>, allowing other threads to impact on the world, and fixing `specs.rs`compatibility.

An option I think about is similar to my Second attempt, with a Trait being a World getter, we might answer all use cases.

Ralith commented:
I don't think specs requires shared ownership, much less explicit locking, of world state.

Vrixyz commented:
For clarification, the World I am putting in Mutex is the one from **nphysics**, not the one from _specs_.

The errors I had concerning `sync::Send` were about putting the `nphysics::World` into `specs::System`.

I'll update the description to make it more explicit.

That being said, you might be right, I didn't find any mentions of `sync::Send` in specs either, but the error messages I had were pretty clear...

For now I focus on making my use case work, then I'll try to optimize the data scheme used.

Thanks for your feedback.

Ralith commented:
You probably encountered the requirements on [`DispatchBuilder::with`](https://docs.rs/specs/*/specs/struct.DispatcherBuilder.html#method.with). These requirements do not apply to all specs systems; indeed you don't need to use dispatchers at all. specs supports mutable data that isn't behind mutexes in many ways, one of which is by storing it as a resource in the specs world.

Vrixyz commented:
That must be it ! Thanks 🙏 Dropping dispatcher for systems requiring nphysics::World might be a solution indeed then :)

Applying Rc<RefCell<nphysics::World> to Testbed is still a requirement though from my understanding.

I’ll make examples from all these informations to help visualizing impacts. *(When I have time)*

Vrixyz commented:
After more testing, this Pull Request kinda works, except for graphics updates. (I added an **Edit 18/08** to first post)

I added this snipper to handle new additions to `nphysics::Wold`, just before drawing.

```rust
for co in physics_world.colliders() {
    if self.graphics.body_nodes_mut(physics_world, co.data().body()).is_none() {
        self.graphics
            .add(window, co.handle(), &physics_world);
        }
}
```

I think that will not handle modifications to existing 2d bodies, but that's a start, I suggest I focus my efforts on supporting new bodies additions during runtime before going in edge cases like modification, deletion, etc.

This Pull request is working at the moment for my (very limited use case).

I'll look into recent informations later to allow :  
- avoid heavy `Arc<Mutex<nphysics::World>>` if not required by (library) user.
- alternate simpler example without `Dispatcher` and its need of `Mutex`.

sebcrozet commented on examples2d/balls2.rs:
```diff
@@ -80,7 +82,7 @@ fn main() {
      * Set up the testbed.
      */
-    let mut testbed = Testbed::new(world);
+    let mut testbed = Testbed::new(Arc::new(Mutex::new(world)));
```
This should be avoided for several reasons:

1.  It means all the testbed examples will have to be modified.
2.  This complicates the example with something that is useful only in some corner-use-cases.
2. I don't want users looking at the examples to think using an `Arc<Mutex<...>>>` is always necessary. (Someone quickly skimming through the examples just to see what they look like will notice the `use` of `Arc` and `Mutex` so they may end up concluding too quickly).

An alternative is to keep the `Testbed::new(world)` as it is, and have a method from the testbed to retrieve the shared world: `testbed.world()`.

Vrixyz replied:
My alternative take on this was to make different instantiaters to pass world to Testbed, from my WIP local branch:

```rust
/// This trait is designed to allow choosing implementation of underlying storing of World: shared between threads or owned only by Testbed.
pub trait WorldOwner {
    fn get<'a: 'b, 'b>(&'a mut self) -> Box<DerefMut<Target = World<f32>> + 'b>;
}

pub struct WorldOwnerShared {
    world: Arc<Mutex<World<f32>>>,
}

impl WorldOwnerShared {
    pub fn new(w: World<f32>) -> WorldOwnerShared {
        WorldOwnerShared {world: Arc::new(Mutex::new(w))}
    }
}

impl WorldOwner for WorldOwnerShared {
    fn get<'a: 'b, 'b>(&'a mut self) -> Box<DerefMut<Target = World<f32>> + 'b> {
        Box::new(self.world.lock().unwrap())
    }
}
// TODO: simpler WorldOwnerExclusive implementation not requiring mutex
```

that would impact my use case example as 
```rust
let mut testbed = Testbed::new(Testbed::WorldOwnerShared(world)));
```
And existing simpler examples as 
```rust
let mut testbed = Testbed::new(Testbed::WorldOwnerExclusive(world)));
```

*Maybe the world itself could implement WorldOwner, so the existing examples wouldn't be impacted at all ? 💃*

Your suggestion makes it difficult or overkill to share `TestBed` with `specs::System`, in order to retrieve `nphysics::World`. **Please tell me** if my lead is incorrect so I don't spend time in a incorrect direction.

All this discussion is underlying to the goal of objective 
>allow to avoid heavy Arc<Mutex<nphysics::World>> if not required by (library) user."

from the **TODO List**

sebcrozet replied:
> Your suggestion makes it difficult or overkill to share TestBed with specs::System, in order to retrieve nphysics::World.

There is no need to share the testbed with specs. What I was thinking of would mean your [example](https://github.com/Vrixyz/testbed-callback/blob/1-arc-mutex/src/main.rs#L122-L136) would look like that:

```rust
    // Initialize the testbed first.
    let mut testbed = Testbed::new(physics_world);

    let ecs_world = std::cell::RefCell::new(specs::World::new());

    // In the following, we have testbed.world() -> Arc<Mutex<World<f32>>
    ecs_world
        .borrow_mut()
        .add_resource(testbed.world());
    let dispatcher = DispatcherBuilder::new()
        .with(DummySystem::new(), "dummy_system", &[])
        .build();

    let dispatcher = std::cell::RefCell::new(dispatcher);

);
```
Basically, the testbed is responsible for creating the `Arc` internally.

-----

If think your solution based on `WorldOwner` is better than what I proposed above. The testbed could be parameterized by: `Testbed<W: WorldOwner>`, and instead of returning `Box<DerefMut<Target = World<f32>> + 'b>>`, the `.get` method you propose could return `impl DerefMut<Target = World<f32>> + 'b>` to avoid boxing.

As you proposed, both `World`, `Arc<RwLock<World>>` and `Arc<Mutex<World>>` should implement this `WorldOwner` trait so no modification of any testbed example is required (even when user callbacks are used since they would become `Fn(&mut T, &mut GraphicsManager, f32)`.

I don't think you will be able to make a solution based on `From<nphysics::World>` without requiring the user to add explicit type annotation for the testbed though.

Vrixyz replied:
What I thought about bypasses modifications of existing examples :tada: : 

```rust
pub fn set_world(&mut self, world: World<f32>) {
    self.set_world_owner(Box::new(WorldOwnerExclusive::from(world)));
}

/// Lower level setter allowing for shared World behaviour
pub fn set_world_owner(&mut self, world: Box<WorldOwner>) ...
```

I tried experiencing with `Into` and `From` for World and `Arc<RwLock>`, but didn't succeed in making the situation better, I'll stop there my experiments on `set_world` API modifications and consider it OK enough.

Vrixyz replied:
I did try to make `get` return a `impl WorldOwned`, but I failed at making it compile again, so I'll let this as an exercise to whoever is more confortable with that.

sebcrozet commented on nphysics_testbed2d/src/testbed.rs:
```diff
@@ -436,25 +441,35 @@ impl State for Testbed {
+                        .add(window, co.handle(), &physics_world);
+                }
+            }
+            self.graphics.draw(&physics_world, window);
```
This was inside of the running check because the `self.graphics.draw(...)` is actually not very well named. All it does is to synchronize the positions of the graphics objects with the position of the physics colliders. Performing this update is not necessary if the simulation is not running.

sebcrozet commented on nphysics_testbed2d/src/testbed.rs:
```diff
@@ -50,11 +52,11 @@ fn usage(exe_name: &str) {
     graphics: GraphicsManager,
     nsteps: usize,
-    callbacks: Vec<Box<Fn(&mut World<f32>, &mut GraphicsManager, f32)>>,
+    callbacks: Vec<Box<Fn(&mut GraphicsManager, f32)>>,
```
I am not very fond of this, though I understand why this can be necessary to avoid dead-locks if, e.g., the user locks the physics world while inside of the callback. Maybe this could at least take `&mut Testbed` as its first argument, so the user can call `testbed.world()` to retrieve the shared world (see my first comment for the definition of `testbed.world()`).

Vrixyz replied:
ok, or pass the `self.world` without locking it, to let the user choose between using the parameter or handling himself the dependency

sebcrozet replied:
Yes, that will work too!

Vrixyz replied:
For reference: passing ` &mut Testbed` in callback caused `cannot reborrow mutably` error and that's why I just plain removed it.

passing the `self.world` needs `Testbed` to be generic over its `WorldOwner` real type. I wanted to avoid adding complexity, but as you suggested it in other comments, I'll look into it.

sebcrozet commented on nphysics_testbed2d/src/testbed.rs:
```diff
@@ -50,11 +52,11 @@ fn usage(exe_name: &str) {
 
 pub struct Testbed {
-    world: World<f32>,
+    world: Arc<Mutex<World<f32>>>,
```
Is there a reason why you used a `Mutex` instead of a `RwLock`? The latter should be more flexible by allowing multiple simultaneous read-only locks.

Vrixyz commented:
@sebcrozet You wrote somewhere that webasm should be tested.

I'm not very comfortable with webassembly and rust, so if you have resources on how to make it work with nphysics, I may follow some guide.

Until then, I think this Pull Request is good for "final" reviews.

Unchecked items from the top TODO List will need more guidance for me to check them. That is, if they are strictly needed, that is for you to decide.

sebcrozet commented:
@Vrixyz Thanks, this PR seems pretty good and almost ready now! I appreciate that your changes don't impact the examples, except for those that fail to compile on travis.

The reason why some examples fail to compile on travis come from the fact the `sensor2` and `body_status2` register callbacks that need to be slightly modified to account for the change of parameter type. For example: https://github.com/sebcrozet/nphysics/blob/master/examples2d/sensor2.rs#L91 should be modified to call the `.get`or `.get_mut` when needed on the world owner.

For testing on WebAssembly, you may follow that tutorial on the user guide: http://nphysics.org/wasm_compatibility/ For testing, you can do this on the `examples2d` repository, and ensure only one of the examples remain on the `Cargo.toml` so it is the only one being compiled.

I will take some time next week to finalize this PR (by performing similar changes to the 3D testbed) and merge it when CI succeeds. If you want, I can do the WebAssembly tests next week as well.

Vrixyz commented:
There's definitely something I'm missing for understanding and testing webasm : when I try to load `target/deploy/index.html`, I have a blank page. Nervermind.

I'm stopping here, CI is green 🎉
